### PR TITLE
Add "Anybody" relation option

### LINKED ISSUE #45: Add "Anybody" option
We claim to create an "all in one" tree, but currently only relatives are included. This options would grow the tree from non-related nodes, i.e. it would include the families of married-in relatives where information exists in the family tree.

This would:

- Add an option "Everyone" next to "Siblings", "Spouses", "Cousins".
- Include all connected individuals, following all parent or child relationships regardless of if they are related.
- The "Siblings", "Spouses", and "Cousins" options would be disabled (greyed out) if "Everyone" is selected.
- Honour the setting "Mark not blood-related people with different color" - i.e. if this setting is enabled and the person is not related to at least one of the individuals, they would use the non-related colour

### PR DESCRIPTION
**Pull request by Neriderc**
This option will include any linked person in the selected generations (respecting max ancestors/descendants settings). In practice, this adds the relatives of spouses.

Resolves #45 and #88

It would be great to have someone test this. It took me a while to get it highlighting the relatives/non-relatives right when option to use different colour is selected. There were also various other bugs I came across so having someone test that this is working right would be great :)

If there's no one available to test I'll just merge into master in a few days and production users can test it :laughing:.

### CODE REVIEW

**Comment by Neriderc** [comment]
Great, thanks! I've just pushed a change that allows overriding the cart into master, and merged it into this so hopefully that doesn't cause any bugs.

**Comment by hartenthaler** [comment]
Sorry for being late. Now I have tested this new function.

It works great. The colors are ok as far as I have seen it. On my screen the dimmed red color for not related women is nearly grey, the dimmed blue color for not related men is ok.

I'm missing partners of partner. For example:
- my sister I. is married to C.
- C. has a son F., a stepson of I.
They are all shown. But F's mother is not included if the option "anyone" is set.

Another example: I'm a widower and my partner G. is a widow. G. has a daughter and a son with her died husband D. All are shown, but not D. So there are some families with only one partner link in the simple mode; in the combined mode it looks very strange (see screenshot).

Screenshot (213)

**Comment by Neriderc** [comment]
Thanks for the feedback! Colours for related/not related are the defaults they have always been, I haven't changed this. You're right though, the pink for non-related is quite difficult to distinguish from the background grey. I'll leave it for now as it's not directly related to what I'm doing here but I think there's a good chance of tweaking it in future.

Thanks for those other scenarios. I am seeing the same thing when I set data up in that way. I would have expected it to work so will need to dig more into it and I'll make some updates, which I'll do before merging this.

Thanks for testing!

**Comment by Neriderc** [comment]
Just as an update on this, I got a little distracted looking at other things but I am still looking at this. I managed to solve the immediate problems but if I try to trace back further it's still not including some relatives of non-relatives. The way the current process is set up makes it more difficult than I expected. I have ideas on this but will need more time. 

It has been a while since I've packaged a release so I'm thinking I'll leave this out for now, grab all the updated translations, give it all a good test, and package a release. Then I'll get back to this for a future release.

**Comment by Neriderc** [comment]
@hartenthaler Could you test this again? I think it's working well... almost too well. Selecting someone and keeping it limited (e.g. 1 ancestor level and 1 descendant) seems to work well, but you can get quite a lot of non-relatives if you have them in your tree! More than a couple of generations led to a huge diagram for me!

From my testing, colouring for relatives/non-relatives seems to work properly, the generations are selected correctly, no one seems to be missing, and I think I've solved the issue of nieces/nephews showing up even when that generation is not supposed to be included (issue #88). Could you see if you can spot any issues?

It's of course also important that the existing settings work, I've done a lot of testing and think it's all working OK, but I have made some changes to how things work so it would be great if you could check if you spot any issues with existing functionality as well.

**Comment by hartenthaler** [comment]
Yes, it is working very well!

I have tested with http://ahnentest.hartenthaler.eu/index.php?route=%2Fmodule%2F_GVExport_%2FChart%2Ftree2&xref=I692 

If I select mode "Simple" and both generation values are set to 0 (zero): I see only the root person. I expect to see his two wives and the husband of his second wife additionally.

If I select mode "Combined" and both generation values are set to 0 (zero): I see the root person and his two wives and a strange box containing only his second wife. I expect to see in this box her and her first husband.

If I change the descendant max level now to 1, I can see the four children of her and her first husband, and now the box is filled correctly with her and her first husband. So everything is ok in this setting.
---
Then I did a bit more complex test: http://ahnentest.hartenthaler.eu/index.php?route=%2Fmodule%2F_GVExport_%2FChart%2Ftree2&xref=I7354 
This is Tina Livanos; using ancestor max level = 0 and descendant level = 1.

First observation: the persons and families that are shown in the Simple mode are very different from the persons and families in the Combined mode. I would expect that the selection of persons and families is the same; only the way to show them should be different. In general: if you select the Combined mode, then the step-parents of a person are not shown, in the Simple mode, you can see step-parents.

Second observation: I would expect to see many more partners of partners in both modes. There are several chains of chains of partners for this person, containing a totally of 31 partners in the same generation (see my custom module "extended family" or use the custom module "extended clippings cart" and then add "this record" and then select "the partner chains Athina Tina Livanos (I7354) belongs to".

Third observation: Eugenia Livanos has two names. Both are shown in her box, but I would expect to see a blank or "/" between both names.

Fourth observation: In the Combined mode there is a box containing Stavros Niarchos married to Charlotte Ford. Under the box there are two strange arrows to "F2847" and "F2848". These are the two marriages of their daughter: Stanley Jozef Olender oo Elena Anne Ford-Niarchos (F2847) and Joseph Daniel Joe Rippolone oo Elena Anne Ford-Niarchos (F2848). I would expect to see two boxes with these partners instead of links to the text "F2847" and "F2848". Such strange links are disappearing sometimes (but not always) if you increase the descendant max level.

**Comment by Neriderc** [comment]
Thanks for testing!


> I have tested with http://ahnentest.hartenthaler.eu/index.php?route=%2Fmodule%2F_GVExport_%2FChart%2Ftree2&xref=I692
> 
> If I select mode "Simple" and both generation values are set to 0 (zero): I see only the root person. I expect to see his two wives and the husband of his second wife additionally.

I wasn't able to get it to run on your link, it just continuously loads when I click the "Update" button.

If I try this with my tree with the original version (first release in this repo) I also only see the root person. I don't think this was caused by this pull request?

> If I select mode "Combined" and both generation values are set to 0 (zero): I see the root person and his two wives and a strange box containing only his second wife. I expect to see in this box her and her first husband.

Hmm if I test in my tree with someone with multiple marriages, I don't see this. Could you test if this works in the latest released version?

> Then I did a bit more complex test: http://
> 
> First observation: the persons and families that are shown in the Simple mode are very different from the persons and families in the Combined mode. I would expect that the selection of persons and families is the same; only the way to show them should be different. In general: if you select the Combined mode, then the step-parents of a person are not shown, in the Simple mode, you can see step-parents.

I haven't changed anything that should change anything between simple/combined mode. There is existing code that changes things a little between the two options but from my understanding it should only change what families are included (i.e. to ensure both partners in a family are shown). I don't think it should affect this. Can you test if it works the same in the last release?

> Second observation: I would expect to see many more partners of partners in both modes. There are several chains of chains of partners for this person, containing a totally of 31 partners in the same generation (see my custom module "extended family" or use the custom module "extended clippings cart" and then add "this record" and then select "the partner chains Athina Tina Livanos (I7354) belongs to".
I wasn't able to access your data but for me, if a family record is included then the partners in the family seem to always be included. However, the siblings of those partners are only included if the ancestor/descendant settings allow that sibling's parent family to be included, in which case siblings are added and then sibling partners etc. Is this a case where the family record isn't added so it doesn't continue with siblings?


> Third observation: Eugenia Livanos has two names. Both are shown in her box, but I would expect to see a blank or "/" between both names.
I'm guessing this is the case with the last release? Names haven't been touched at all in this work.

> Fourth observation: In the Combined mode there is a box containing Stavros Niarchos married to Charlotte Ford. Under the box there are two strange arrows to "F2847" and "F2848". These are the two marriages of their daughter: Stanley Jozef Olender oo Elena Anne Ford-Niarchos (F2847) and Joseph Daniel Joe Rippolone oo Elena Anne Ford-Niarchos (F2848). I would expect to see two boxes with these partners instead of links to the text "F2847" and "F2848". Such strange links are disappearing sometimes (but not always) if you increase the descendant max level.
Is this something you only see in this version?

All in all I'm keen to know what is new from this pull request and what may have been like that already. There are probably a bunch of issues to open based on this testing, but if they are existing then I don't think it should stop this pull request being merged. We can fix it in the future.

I wasn't able to view your test webtrees environment which has made it a little hard to follow what the issues are. It would be good to see what's new and existing then I will work on trying to reproduce using test data on my instance. By the way, thanks very much for testing!

**Comment by hartenthaler** [comment]
I just logged out from my webtrees test installation and tested again http://ahnentest.hartenthaler.eu/index.php?route=%2Fmodule%2F_GVExport_%2FChart%2Ftree2&xref=I692 (selecting I692 as starting person). It loads after two seconds. Nobody else than me is using this site, but it should work for you. Strange. Did you observe a problem only during using GVExport or in general when accessing this test site?

Testing with the latest released version: in the combined mode, I can see two boxes: the root person and his first and his second wife.; this is as expected because "partner" is selected. Switching to Simple mode, shows only the root person. This is maybe as it was always, but I would expect to see three persons and two family elipses, the root person, his two wives and their families.

The boxes with only one partner in Combined mode are only visible in the any_option code, not in the last released version.
Screenshot (217)

> I haven't changed anything that should change anything between simple/combined mode. There is existing code that changes things a little between the two options but from my understanding it should only change what families are included (i.e. to ensure both partners in a family are shown). I don't think it should affect this. Can you test if it works the same in the last release?

Yes, I had looked at the code when preparing the function to add the content of the clippings cart to the internal data structures. In the latest released version there works everything as before (beside my new expectation as explained above).

The problem with two names happens in the latest released code, too. It happens only in the combination: combined mode with graph direction = top-to-bottom. So this is a separate issue.
Screenshot (218)

I have seen links to families as text instead of links to family elipses already before, during testing my additional function for the clippings cart.
Screenshot (219)

**Comment by Neriderc** [comment]
Today I have tried again and your site works fine. Not sure what the problem was yesterday.

I've had a bit of a play and have for the most part seen the issues you described. I haven't been able to reproduce all issues using data from my own tree that I thought was comparable, but largely it seems to me these issues already exist in the current version and are not related to this pull request.We should try to fix these, but they can be looked at separately.

I think the only issue directly related to this pull request may be your mention of not seeing all partners of partners? Is this just a difference in expectation of which records are included? For example, if you include a generation of ancestors, only the parent's of the starting individual are included (i.e. their parents, and parents of any adopted siblings). If you want siblings of the parents to be included, their family record with the parents must be included, which requires going back another generation. Does that explain the difference or is there still a problem with partners?

Using your site I don't seem to be able to access the extended clippings cart module so haven't been able to follow your directions to have this demonstrated.

**Comment by hartenthaler** [comment]
The extended clippings cart is now visible for visitors (for the tree that is used in the links above).

Yes, the only issues related to the new code are
1 sometimes there are in the combined mode boxes with only one partner, even if the other partner exists in the tree.
2 if there are chains of partners of partners, I would expect to see them all because they are all in the same generation as the starting person. For example when selecting Athina http://ahnentest.hartenthaler.eu/index.php?route=%2Fmodule%2F_GVExport_%2FChart%2Ftree2&xref=I7354 then there should be 31 partners of partners (see http://ahnentest.hartenthaler.eu/index.php?route=%2Ftree%2Ftree2%2Findividual%2FI7354%2FAthina-Livanos#tab-_hh_extended_family_). I assume that this issue is a result of issue 1, because if you do not find the partner of a person (box with only one partner), then you will not find the additional partners of the so missing person.
3 sometimes there are links to a family using text ("F2847") instead of using a box with the child and his/her partner

**Comment by Neriderc** [comment]
I think 1 and 3 are not related to this pull request. I have definitely reproduced 3 using older releases of GVExport, and I think 1 is also an existing issue that I think should be solved separately. i.e. if you compare a previous version of GVExport against this version, with "Anyone" _unchecked_, then they should display the same thing.

> 2 if there are chains of partners of partners, I would expect to see them all because they are all in the same generation as the starting person. 
In theory this sounds right, but because of how GVExport works it won't always be the case. If a single individual is listed with no descendants or ancestors, I'd expect no one else listed. This is because the individual is not linked to an ancestor family where e.g. their siblings would be found. The individual is also not linked to their descendant family, where their children and partners would be found.

While I'm open to changing this with a good enough argument, I don't want to make this change as part of this piece of work. This pull request is specifically about following the existing way of mapping the tree, but not stopping just because a found individual is not related.

> For example when selecting Athina http://ahnentest.hartenthaler.eu/index.php?route=%2Fmodule%2F_GVExport_%2FChart%2Ftree2&xref=I7354 then there should be 31 partners of partners (see http://ahnentest.hartenthaler.eu/index.php?route=%2Ftree%2Ftree2%2Findividual%2FI7354%2FAthina-Livanos#tab-_hh_extended_family_). I assume that this issue is a result of issue 1, because if you do not find the partner of a person (box with only one partner), then you will not find the additional partners of the so missing person.

I do see the partner chain should be 31 and that not all those are included. However, most show as "Private" and in any case it's difficult to debug when it's a different server and difficult to know if it's due to issue 1, the one I could see details for made me think perhaps this is not the case. Especially since the issue exists outside of combined mode. Is there a small, straightforward example that you know causes the issue? If I add a test tree and start chaining up partners (add ex wife, add partner of ex wife, add ex partner or partner of ex wife, etc), they are all shown as I expect.

I think without a small and specific testable example it's hard to work out the issue. I have been thinking this project could do with a test GEDCOM, one that is designed to have a list of specific scenarios in the tree so that we can test against them. If we find an issue not in the test GEDCOM, we can add it. That way different people can test from the same base data for easy reproducibility.

**Comment by Neriderc** [comment]
It seems partners of partners are not included in your server. For example, if I choose Aristoteles as the starting individual, then his wife Jacqueline is added, great. However, her first husband John is not shown. But, the family record for the Jacqueine/John _is_ included. 
image

But when I try to create this scenario on my server,

**Comment by hartenthaler** [comment]
Maybe you are right that these issues are unrelated to the code of any_option. I have produced a GEDCOM test file using the chain of partners using the extended clippings cart, added the chain to the clippings cart, deleted everything besides the INDI+FAM records, and exported the result as a GEDCOM file (it is attached as .txt, so you have to change .txt to .ged).

When using this for testing GVExport I found that in the simple mode if both max generations are set to 0, then only the starting person is shown. When selecting the max generations to 1 (what should not change anything, because there are only persons in generation #0 in the file, no ancestors, no descendants), then in simple mode there are at least some persons shown (as it is in the combined mode).

[wtcce(I7354).txt](https://github.com/Neriderc/GVExport/files/9240606/wtcce.I7354.txt)

**Comment by Neriderc** [comment]
Thanks! I expect that no others would be shown when setting generations to 0, this is just how it's set up currently (as family records are not included therefore no links to follow).

However, I definitely see that partners of partners are not included. I will use this data to work on tracking down why this is. It's very helpful, thanks!

**Comment by hartenthaler** [comment]
> However, her first husband John is not shown.

What is the specialty for John compared to the ex-other-husband in your example? John is dead, so it should not be related to security.

**Comment by Neriderc** [comment]
I've just pushed a small change, could you see if this fixes it?

I think if there were kids then their parents were added and therefore the partners added. In this case there were no kids and so there wasn't a backup link, so the partner wasn't added.

I've just changed the recursion so it follows the spouse link after adding a spouse. Before it made sense not to because you didn't get this situation, but now I've enabled following this link when the "Anyone" option is checked.

**Comment by Neriderc** [comment]
I have also managed to reproduce the family in combined mode showing as just the family XREF, and in this I have it only happening when "Anyone" is selected. So I will expend some more effort to look that.

From my initial look, the output array is *exactly* the same in my test data whether "Anyone" is checked or not (used data that doesn't have extended non-relatives), yet it works when "Anyone" is not selected, and doesn't work when it is selected. And I didn't think anything outside this array generation has been affected, so it's certainly a puzzler.

**Comment by Neriderc** [comment]
Right, have just pushed another small change to resolve the issue with the family sometimes showing as only the XREF in combined mode when "Anyone" option is enabled. This was due to an array being overwritten the second time a record was checked - which is possible in specific circumstances. I've made a change to only update the array value if it wasn't already set, and this appears to resolve the issue.

As you've seen I've created #101 to discuss your issue 1, so I believe that all known issues on this pull request have been addressed?

I've very keen to get this merged since there are a growing number of things to work on and I didn't want to start something new and risk not being able to merge because of a conflict :grin:.

**Comment by Neriderc** [comment]
Merging this to master so I can progress some other things. Let me know if you have issues, we can correct things before we release another bundled version.

**Comment by hartenthaler** [comment]
Super! I will test it!

**Comment by Neriderc** [comment]
Awesome, thanks!

**Comment by hartenthaler** [comment]
I just tested against the latest development code. Great work! Thank you very much! It is now much better. The chains of partners of partners are now complete. Underscoring of names marked by an asterisk is working now. I found only some less important things that are already open issues. So it would be great if you can release a new version in the next time.

**Comment by Neriderc** [comment]
Great, over the next day or so I'll look at giving the master branch a good testing to make sure it's ok, and will update the translation files with the latest ones from POEditor, then will release an updated version.
